### ISSUE TITLE
New Rule Suggestion: no-inferred-method-name

### ISSUE THREAD

**Issue by ericelliott** [eslint/eslint#2454]
Every form of function expression assignment in ES6 infers a `name` that can be used inside the function for recursion, *except* object literal assignments. With proper tail calls, any call to that name in tail position will optimize to a while loop, which is very fast and attractive.

Therefore, **it's easy to assume** that concise methods in object literals behave the same way. However, **that is not true.**

```js
const a = {
  foo() {
  }
}
```

**is not equivalent to:**

```js
const a = {
  foo: function foo () {
  }
}
```

It's more like this (in ES5):

```js
const a = {
  foo: function () {
  }
}
```

A more detailed example:

```js
const repeater = {
  name: 'Repeater',
  types: [
    { f: 'function' },
    { n: 'number' }
  ],
  repeat (f, n) {
    let i = n - 1;

    if (typeof f === 'function') {
      f();
    } else {
      throw new Error('repeat: A Function is required.');
    }
    if (!n) {
      return undefined;
    }

    return repeat(f, i);
  }
};

repeater.repeat(() => {console.log('foo');}, 10);
```

Produces this output:
```
foo
ReferenceError: repeat is not defined
```

The same thing happens if you instead use:

```js
    repeat: function (f, n) {
```

"repeat" gets assigned to `repeater.repeat.name`, but `repeat` is undefined inside the function at runtime. Very confusing.

But this works:

```js
    repeat: function repeat (f, n) {
```

In other words, it logs "foo" to the console until `!n` evaluates to true.

This issue has already tripped up several people I know of, including me and Kyle Simpson, who wrote a book on the topic of the ES6 spec. Very frustrating!

The workaround:

1. For recursion in methods, always use an explicit **named function expression**: `repeat: function repeat (f, n) {`. Remember that `this` does not provide a reliable reference to the recursive function like a named function expression does.

For reference and a painfully hand-held tour through the ES6 spec RE: this issue - https://github.com/babel/babel/issues/1367

**Comment by ericelliott** [comment]
Note: `no-undef` catches this error, but doesn't explain why it's undefined or how to fix it.

**Comment by michaelficarra** [comment]
I don't think the `no-undef` rule should try to explain *why* a reference doesn't exist. Maybe you just forgot a `let`. :-1:

**Comment by ericelliott** [comment]
I don't think the `no-undef` rule should explain `why` either. That's why I want this `no-inferred-method-name` rule, so `no-undef` can just throw its hands up in the air, but the developer still gets a more useful error that can actually help the developer solve the problem.

**Comment by nzakas** [comment]
Given that this case is already properly flagged by `no-undef`, I don't think there's a place for this rule. We don't let rules overlap their warnings as it is confusing, and I can't see a compelling reason to make `no-undef` skip this case.

You are, of course, free to create a custom rule.

**Comment by ericelliott** [comment]
Now there's a [custom ESLint rule](https://github.com/johnstonbl01/eslint-no-inferred-method-name) for this issue. Thanks for the discussion. =)

And thanks for making ESLint customizable. That is a very useful feature.

Cheers!
